src/scope.rs

```diff
@@ -1,292 +1,961 @@
-// Copyright 2019-2020 the Deno authors. All rights reserved. MIT license.
-
+use std::any::type_name;
+use std::cell::Cell;
 use std::marker::PhantomData;
+use std::mem::replace;
 use std::mem::size_of;
-use std::mem::take;
 use std::mem::MaybeUninit;
+use std::num::NonZeroUsize;
+use std::ops::Deref;
+use std::ops::DerefMut;
+use std::ptr;
+use std::ptr::null;
 use std::ptr::NonNull;
 
-use crate::scope_traits::internal::GetRawIsolate;
+use crate::function::FunctionCallbackInfo;
+use crate::function::PropertyCallbackInfo;
 use crate::Context;
-use crate::FunctionCallbackInfo;
-use crate::InIsolate;
+use crate::Data;
 use crate::Isolate;
 use crate::Local;
+use crate::Message;
+use crate::Object;
+use crate::OwnedIsolate;
+use crate::Primitive;
 use crate::PromiseRejectMessage;
-use crate::PropertyCallbackInfo;
-
-// Note: the 's lifetime is there to ensure that after entering a scope once,
-// the same scope object can't ever be entered again.
-
-/// A trait for defining scoped objects.
-pub unsafe trait ScopeDefinition<'s>
-where
-  Self: Sized,
-{
-  type Args;
-  unsafe fn enter_scope(buf: *mut Self, args: Self::Args);
-}
+use crate::TryCatch;
+use crate::Value;
+
+#[doc(inline)]
+pub use api::{CallbackScope, ContextScope, EscapableHandleScope, HandleScope};
+pub(crate) use data::ScopeData;
+
+mod api {
```

Comment:
The `api` and `data` modules presumably are some sort of organizational structure - but it's not obvious to me what they mean exactly - it would help readers to have a small triple-slash description at the top of each.